cmd/experimental/kjobctl/pkg/builder/slurm_builder.go

```diff
@@ -0,0 +1,647 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package builder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math"
+	"os"
+	"slices"
+	"strconv"
+	"strings"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilrand "k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/utils/ptr"
+
+	"sigs.k8s.io/kueue/cmd/experimental/kjobctl/apis/v1alpha1"
+	"sigs.k8s.io/kueue/cmd/experimental/kjobctl/pkg/parser"
+)
+
+const (
+	slurmScriptFilename     = "script.sh"
+	slurmEntrypointFilename = "entrypoint.sh"
+	slurmPath               = "/slurm"
+
+	//# \\ - Do not process any of the replacement symbols.
+	//# %% - The character "%".
+	//# %A - Job array's master job allocation number (for now it is equivalent to SLURM_JOB_ID).
+	//# %a - Job array ID (index) number (SLURM_ARRAY_TASK_ID).
+	//# %j - job id of the running job (SLURM_JOB_ID).
+	//# %N - short hostname (pod name).
+	//# %n - node(pod) identifier relative to current job - index from K8S index job.
+	//# %t - task identifier (rank) relative to current job - It is array id position.
+	//# %u - username (from the client machine).
+	//# %x - job name.
+	unmaskFilenameFunction = `unmask_filename () {
```

Comment:
what is this function doing?

Reply:
It is for replacing %-variables on the filename. Look [here](https://slurm.schedmd.com/sbatch.html#SECTION_FILENAME-PATTERN).